src/main/java/net/snowflake/ingest/streaming/internal/IcebergParquetValueParser.java

```diff
@@ -502,35 +494,34 @@ private static ParquetBufferValue get3LevelMapValue(
       Map<String, RowBufferStats> statsMap,
       ZoneId defaultTimezone,
       final long insertRowsCurrIndex,
-      String path,
-      boolean isDescendantsOfRepeatingGroup) {
+      String path) {
     Map<?, ?> mapVal =
         DataValidationUtil.validateAndParseIcebergMap(path, value, insertRowsCurrIndex);
     List<Object> listVal = new ArrayList<>();
-    final AtomicReference<Float> estimatedParquetSize = new AtomicReference<>(0f);
-    mapVal.forEach(
-        (k, v) -> {
-          ParquetBufferValue parsedKey =
-              parseColumnValueToParquet(
-                  k,
-                  type.getType(0).asGroupType().getType(0),
-                  statsMap,
-                  defaultTimezone,
-                  insertRowsCurrIndex,
-                  concatDotPath(path, THREE_LEVEL_MAP_GROUP_NAME),
-                  true);
-          ParquetBufferValue parsedValue =
-              parseColumnValueToParquet(
-                  v,
-                  type.getType(0).asGroupType().getType(1),
-                  statsMap,
-                  defaultTimezone,
-                  insertRowsCurrIndex,
-                  concatDotPath(path, THREE_LEVEL_MAP_GROUP_NAME),
-                  isDescendantsOfRepeatingGroup);
-          listVal.add(Arrays.asList(parsedKey.getValue(), parsedValue.getValue()));
-          estimatedParquetSize.updateAndGet(sz -> sz + parsedKey.getSize() + parsedValue.getSize());
-        });
-    return new ParquetBufferValue(listVal, estimatedParquetSize.get());
+    float estimatedParquetSize = 0;
+    String mapGroupPath = concatDotPath(path, THREE_LEVEL_MAP_GROUP_NAME);
+    for (Map.Entry<?, ?> entry : mapVal.entrySet()) {
+      ParquetBufferValue parsedKey =
+          parseColumnValueToParquet(
+              entry.getKey(),
+              type.getType(0).asGroupType().getType(0),
+              statsMap,
+              defaultTimezone,
+              insertRowsCurrIndex,
+              mapGroupPath,
+              true);
+      ParquetBufferValue parsedValue =
+          parseColumnValueToParquet(
+              entry.getValue(),
+              type.getType(0).asGroupType().getType(1),
+              statsMap,
+              defaultTimezone,
+              insertRowsCurrIndex,
+              mapGroupPath,
+              true);
+      listVal.add(Arrays.asList(parsedKey.getValue(), parsedValue.getValue()));
+      estimatedParquetSize += parsedKey.getSize() + parsedValue.getSize();
```

Comment:
Is there no overhead bytes that need to be added in this case? So the parquet file adds a byte[] for the key, and byte[] for the value, but (a) no separator, or (b) no header that writes out the number of bytes of each field?

How are we validating that the estimatedparquetSize is accurate with structured data types?

Reply:
Afaik parquet use definition & repetition level to determine the structure. The headers only exist for encoding (e.g. bitpack encoding needs 1 bytes header at the start of the run). For estimated parquet size, we only validates this through our assertion test, do you think we should add a test case against the actual generated blob?

Reply:
yeah lets add this to the testing gaps JIRA, to veirfy how close our estimate gets to actual file size.

Reply:
Added in [jira](https://snowflakecomputing.atlassian.net/browse/SNOW-1490151).